src/main/java/com/nvidia/spark/rapids/jni/HyperLogLogPlusPlusHostUDF.java

```diff
@@ -0,0 +1,142 @@
+  }
+
+  /**
+   * HyperLogLogPlusPlus(HLLPP) aggregation types
+   */
+  public enum AggregationType {
+
+    /**
+     * Compute hash codes for the input, generate HyperLogLogPlusPlus(HLLPP)
+     * sketches from hash codes, and merge all the sketches into one sketch, output
+     * is a struct scalar with multiple long values.
+     */
+    Reduction(0),
+
+    /**
+     * Merge all HyperLogLogPlusPlus(HLLPP) sketches in the input column into one
+     * sketch. Input is a struct column with multiple long columns which is
+     * consistent with Spark. Output is a struct scalar with multiple long values.
+     */
+    ReductionMERGE(1),
+
+    /**
+     * Compute hash codes for the input, generate HyperLogLogPlusPlus(HLLPP)
+     * sketches from hash codes, and merge the sketches in the same group. Output is
+     * a struct column with multiple long columns which is consistent with Spark.
+     */
+    GroupBy(2),
+
+    /**
+     * Merge HyperLogLogPlusPlus(HLLPP) sketches in the same group.
+     * Input is a struct column with multiple long columns which is consistent with
+     * Spark.
+     */
+    GroupByMerge(3);
+
+    final int nativeId;
+
+    AggregationType(int nativeId) {
+      this.nativeId = nativeId;
+    }
+  }
+
+  /**
+   * Create a HyperLogLogPlusPlus(HLLPP) host UDF
+   */
+  private static long createHLLPPHostUDF(AggregationType type, int precision) {
+    return createHLLPPHostUDF(type.nativeId, precision);
+  }
+
+  /**
+   * Compute the approximate count distinct value from sketch values.
+   *
+   * The input is sketch values, must be given in the format:
+   * `Struct<INT64, INT64,...>`,
+   * The value of num_registers_per_sketch = 2^precision
+   * The children num of this Struct is: num_registers_per_sketch / 10 + 1,
+   * Here 10 means a INT64 contains 10 register values,
+   * each register value is 6 bits.
+   * Register value is the number of leading zero bits in xxhash64 hash code.
+   * xxhash64 hash code is 64 bits, Register value is 6 bits,
+   * 6 bits is enough to hold the max value 64.
+   *
+   * @param input     The sketch column which constains Struct<INT64, INT64, ...>
+   *                  values.
+   * @param precision The num of bits for HLLPP register addressing.
+   * @return A INT64 column with each value indicates the approximate count
+   *         distinct value.
+   */
+  public static ColumnVector estimateDistinctValueFromSketches(ColumnView input, int precision) {
+    return new ColumnVector(estimateDistinctValueFromSketches(input.getNativeView(), precision));
+  }
+
+  private static native long createHLLPPHostUDF(int type, int precision);
+
+  private static native long estimateDistinctValueFromSketches(long inputHandle, int precision);
+
+  private AggregationType type;
+  private int precision;
+
+  /**
+   * TODO: move this to cuDF HostUDFWrapper
+   */
+  @Override
+  public void close() throws Exception {
+    close(udfNativeHandle);
+  }
```

Comment:
If we move `close()` to cudf `HostUDFWrapper`, we have to declare another abstract non-static method like `closeUDF(long nativePtr)` or similar which must be implemented in this class. And such non-static method here will call to the static method `close(long ptr)`, adding more complexity to the overall implementation.